src/dist/replication/lib/mutation.cpp

```diff
@@ -192,25 +168,46 @@ void mutation::write_to_log_file(std::function<void(const blob&)> inserter) cons
     }
 }
 
-void mutation::write_to_log_file(binary_writer& writer) const
+void mutation::write_to(binary_writer& writer, dsn_message_t to) const
 {
     writer.write_pod(data.header);
     writer.write_pod(static_cast<int>(data.updates.size()));
 
     for (const mutation_update& update : data.updates)
     {
-        writer.write_pod(static_cast<int>(update.code));
+        // write task_code as string to make it cross-process compatible.
+        // avoid memory copy, equal to writer.write(std::string)
+        const char* cstr  = update.code.to_string();
+        int len = static_cast<int>(strlen(cstr));
+        writer.write_pod(len);
+        if (len > 0)
+            writer.write(cstr, len);
+
         writer.write_pod(static_cast<int>(update.serialization_type));
+
         writer.write_pod(static_cast<int>(update.data.length()));
     }
 
-    for (const mutation_update& update : data.updates)
+    if (nullptr != to)
     {
-        writer.write(update.data.data(), update.data.length());
+        writer.flush();
+
+        message_ex* msg = (message_ex*)to;
```

Comment:
```message_ex``` is not exposed to upper apps and frameworks, which invokes the core features through C api eventually.